others/lru_cache.cpp

```diff
@@ -0,0 +1,269 @@
+ * the cache and update the address in hash map for that page.
+ *
+ * @author [Nitin Sharma](https://github.com/foo290)
+ * */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for IO Operations
+#include <list>           /// for std::list
+#include <unordered_map>  /// for std::unordered_map
+
+/**
+ * @namespace others
+ * @brief Other algorithms
+ */
+namespace others {
+/**
+ * @namespace lru_cache
+ * @brief Implementation of LRU caching algorithm
+ */
+namespace lru_cache {
+
+/**
+ * @brief LRU cache class
+ */
+class LRUCache {
+    uint64_t pageFrame;         ///< Page frame, or total size of the cache.
+    std::list<uint64_t> cache;  ///< Cache linked list (using the STL)
+    std::unordered_map<uint64_t, std::list<uint64_t>::iterator>
+        pageMap;  ///< Hash map containing pages and their addresses
+
+    uint64_t hits =
+        0;  ///< Total number of hits, or total number of times a page
+    ///< was found in cache.
+    uint64_t pageFault =
+        0;  ///< Total number of miss/page fault, or total number of
+            ///< times a page was not found in cache
+
+ public:
+    /**
+     * @brief Constructor, Initialize thee LRU class with page frame.
+     * @param pf Page frame or total size of cache.
+     * */
+    explicit LRUCache(uint64_t pf) { pageFrame = pf; }
+
+    /**
+     * @brief Refer to a page, or request a page from memory.
+     * @param page The page that you are referring to.
+     * @returns void
+     * */
+    void refer(uint64_t page) {
+        // If the page requested not in cache.
+        if (pageMap.find(page) == pageMap.end()) {
+            pageFault++;  ///< Increase the page fault by one.
+
+            // Check if the cache is full
+            if (cache.size() == pageFrame) {
+                // delete the last page from cache
+                uint64_t lastPage = cache.back();
+                cache.pop_back();
+                pageMap.erase(lastPage);
+            }
+        }
+        // The requested page is in the cache
+        else {
+            hits++;
+            // present in cache, erase from current position to bring in front
+            cache.erase(pageMap[page]);
+        }
+        // Push it in the front of the cache and update the page reference in
+        // page map.
+        cache.push_front(page);
+        pageMap[page] = cache.begin();
+    }
+
+    /**
+     * @brief A function to display the current cache
+     * @returns Void
+     * */
+    void display() {
+        for (uint64_t &it : cache) {
+            std::cout << it << " ";
+        }
+        std::cout << std::endl;
+    }
+    /**
+     * @brief A function to get page hits
+     * @returns int
+     * */
+    uint64_t getHits() const { return hits; }
+    /**
+     * @brief A function to get page fault
+     * @returns int
+     * */
+    uint64_t getPageFault() const { return pageFault; }
+};
+
+}  // namespace lru_cache
+}  // namespace others
```

Comment:
```suggestion
}  // namespace others
}  // namespace lru_cache
```

Reply:
I think this is right, it's a clang-tidy fix, the first namespace is others and the second is lru_cache, so lru_cache will be closed before others.

Reply:
Ohh, I got confused when I made that review. Sorry about that. 😅